### DISCUSSION TITLE
How to migrate from an existing express project?

### DISCUSSION THREAD

**Discussion by hunght** [RobinTail/express-zod-api#647] [Q&A] [answered]
- As the question. we have an existing express project with a lot of things going on, what is a strategy to do it little by little cause we can't have the effort to do it one time? Thanks for the awesome works

**Comment by RobinTail** [comment, marked as answer]
Hello @hunght ,

I'd recommend the following migration strategy.

1. Install `express-zod-api` and use `attachRouting()` method with your existing express `app` as shown in the following example:
  https://github.com/RobinTail/express-zod-api#connect-to-your-own-express-app
2. Make sure that it still works with an empty `routing` object, consider replacement of the your 404-handler (if any) to the one returned by `attachRouting()`.
3. Here begins the iterative process of migrating your endpoints one by one.
  a. Use the [documentation](https://github.com/RobinTail/express-zod-api#create-your-first-endpoint) and [examples](https://github.com/RobinTail/express-zod-api/tree/master/example) to create a copy of one of your endpoint handler using `express-zod-api` methods, factories, endpoints builders, middlewares and so on.
  b. Remove the pointer of your original endpoint from your original express routing.
  c. Add the pointer of your newly created endpoint to the `routing` object used by `attachRouting()`.
  d. Make sure that the new copy is now operating.
  e. Make sure that it works as expected (run your tests if any).
  f. Remove the original implementation of the endpoint. Now it has been migrated to `express-zod-api`.
4. When all endpoints migrated, consider switching to `createServer()` instead of `attachRouting` in case it turned out that your app became simpler.

**Reply by hunght** [reply]
Thank you! @RobinTail. Now it is clear to me to apply then.

**Reply by hunght** [reply]
one more question @RobinTail, How do we silence all validated errors (input and output)? We want to keep it in the logger only. not throw exceptions so we can resolve it first before forcing it

**Reply by RobinTail** [reply]
> one more question @RobinTail, How do we silence all validated errors (input and output)? We want to keep it in the logger only. not throw exceptions so we can resolve it first before forcing it 

Those errors are the whole point of using the library – it prevents receiving unexpected values into variables your endpoints operate.
There is no way to replace validation errors with warnings.
However, you can avoid errors by creating more loose I/O schemas.
Instead of `z.string()` you can do `z.any()` for example. You can also use `.or()`.  @hunght

**Reply by hunght** [reply]
Yes, I got that. But somehow the dev makes a miss take in which the input can be both string or number not only number then the API fails which we do not want. Or the output can be null instead of undefined. I could try to make the PR for make optional on that if you agree. @RobinTail

**Reply by hunght** [reply]
the idea here is we want to review the issue first to see if it is real problem or just a wrong alert as the system is already working, and the type-safe in our app is not very strict. We convert from js code to typescript and `any` keyword is everywhere

**Reply by hunght** [reply]
Hi @RobinTail  I made a PR for that https://github.com/RobinTail/express-zod-api/pull/652 pls help check. Thank you

**Reply by RobinTail** [reply]
> and `any` keyword is everywhere

@hunght , thank you for your efforts, but the problem you're trying to fix is not within the library.

**Reply by hunght** [reply]
okay thank for take time on this @RobinTail

**Reply by elmeister** [reply]
@RobinTail I believe we have the same issue and it is also mentioned in https://github.com/RobinTail/express-zod-api/discussions/695.
The impact of having not working API endpoint is way more problematic than incorrect type in it, an option to be able to still proceed with response is critical for us.

**Reply by RobinTail** [reply]
@elmeister , https://github.com/RobinTail/express-zod-api/discussions/695#discussioncomment-4369164
